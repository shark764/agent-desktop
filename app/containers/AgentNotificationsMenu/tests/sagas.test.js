/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import {
  goInitializeNotificatonPreferences,
  goToggleAudioNotificationsPreference,
  goToggleVisualNotificationsPreference,
} from '../sagas';

describe('goInitializeNotificatonPreferences', () => {
  afterAll(() => {
    delete global.localStorage;
  });
  describe('notifications stored in localStorage', () => {
    const mockGetItem = jest.fn().mockReturnValue('true');
    beforeEach(() => {
      global.localStorage = {
        getItem: mockGetItem,
      };
    });
    const generator = goInitializeNotificatonPreferences();
    it('selectes the tenant id', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('selectes the agent id', () => {
      expect(generator.next({ id: 'tenantId' })).toMatchSnapshot();
    });
    it('puts the audio notification into state', () => {
      expect(generator.next({ userId: 'agentId' })).toMatchSnapshot();
    });
    it('puts the visual notification into state', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('uses the tenant and agent id to get the items from localStorage', () => {
      expect(mockGetItem.mock.calls).toMatchSnapshot();
    });
    it('is done', () => {
      expect(generator.next().done).toBe(true);
    });
  });
  describe('notifications not stored in localStorage', () => {
    beforeEach(() => {
      global.localStorage = {
        getItem: () => null,
      };
    });
    const generator = goInitializeNotificatonPreferences();
    it('selectes the tenant id', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('selectes the agent id', () => {
      expect(generator.next({ id: 'tenantId' })).toMatchSnapshot();
    });
    it('is done', () => {
      expect(generator.next({ userId: 'agentId' }).done).toBe(true);
    });
  });
});

describe('goToggleAudioNotificationsPreference', () => {
  const mockSetItem = jest.fn();
  beforeEach(() => {
    global.localStorage = {
      setItem: mockSetItem,
    };
  });
  afterAll(() => {
    delete global.localStorage;
  });
  const generator = goToggleAudioNotificationsPreference();
  it('gets the current audio notification', () => {
    expect(generator.next()).toMatchSnapshot();
  });
  it('puts the inverse of it', () => {
    expect(generator.next(true)).toMatchSnapshot();
  });
  it('selectes the tenant id', () => {
    expect(generator.next()).toMatchSnapshot();
  });
  it('selectes the agent id', () => {
    expect(generator.next({ id: 'tenantId' })).toMatchSnapshot();
  });
  it('is done', () => {
    expect(generator.next({ userId: 'agentId' }).done).toBe(true);
  });
  it('uses the tenant and agent id to set the notification in localStorage', () => {
    expect(mockSetItem.mock.calls).toMatchSnapshot();
  });
});

describe('goToggleVisualNotificationsPreference', () => {
  afterAll(() => {
    delete global.localStorage;
    delete global.Notification;
  });
  describe('does not have notification permission', () => {
    const mockSetItem = jest.fn();
    beforeEach(() => {
      global.localStorage = {
        setItem: mockSetItem,
      };
      global.Notification = {
        permission: 'denied',
        requestPermission: () => new Promise((resolve) => resolve(true)),
      };
    });
    describe('permission is granted upon request', () => {
      const generator = goToggleVisualNotificationsPreference();
      it('requests permission', () => {
        expect(generator.next()).toMatchSnapshot();
      });
      it('sets the notification to true', () => {
        expect(generator.next('granted')).toMatchSnapshot();
      });
      it('selectes the tenant id', () => {
        expect(generator.next()).toMatchSnapshot();
      });
      it('selectes the agent id', () => {
        expect(generator.next({ id: 'tenantId' })).toMatchSnapshot();
      });
      it('is done', () => {
        expect(generator.next({ userId: 'agentId' }).done).toBe(true);
      });
      it('uses the tenant and agent id to set the notification in localStorage', () => {
        expect(mockSetItem.mock.calls).toMatchSnapshot();
      });
    });
    describe('permission is not granted upon request', () => {
      const generator = goToggleVisualNotificationsPreference();
      it('requests permission', () => {
        expect(generator.next()).toMatchSnapshot();
      });
      it('adds error', () => {
        expect(generator.next('denied')).toMatchSnapshot();
      });
      it('is done', () => {
        expect(generator.next().done).toBe(true);
      });
    });
  });
  describe('does have notification permission', () => {
    const mockSetItem = jest.fn();
    beforeEach(() => {
      global.localStorage = {
        setItem: mockSetItem,
      };
      global.Notification = {
        permission: 'granted',
      };
    });
    const generator = goToggleVisualNotificationsPreference();
    it('gets the current visual notification', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('puts the inverse of it', () => {
      expect(generator.next(true)).toMatchSnapshot();
    });
    it('selectes the tenant id', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('selectes the agent id', () => {
      expect(generator.next({ id: 'tenantId' })).toMatchSnapshot();
    });
    it('is done', () => {
      expect(generator.next({ userId: 'agentId' }).done).toBe(true);
    });
    it('uses the tenant and agent id to set the notification in localStorage', () => {
      expect(mockSetItem.mock.calls).toMatchSnapshot();
    });
  });
});
