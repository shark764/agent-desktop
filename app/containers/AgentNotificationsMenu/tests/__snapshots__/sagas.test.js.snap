// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`goInitializeNotificatonPreferences notifications not stored in localStorage selectes the agent id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications not stored in localStorage selectes the tenant id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications stored in localStorage puts the audio notification into state 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "audioNotifications": true,
        "type": "app/AgentNotificationsMenu/SET_AUDIO_NOTIFICATIONS_PREFERENCE",
      },
      "channel": null,
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications stored in localStorage puts the visual notification into state 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/AgentNotificationsMenu/SET_VISUAL_NOTIFICATIONS_PREFERENCE",
        "visualNotifications": true,
      },
      "channel": null,
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications stored in localStorage selectes the agent id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications stored in localStorage selectes the tenant id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goInitializeNotificatonPreferences notifications stored in localStorage uses the tenant and agent id to get the items from localStorage 1`] = `
Array [
  Array [
    "skylightAudioNotifications-tenantId-agentId",
  ],
  Array [
    "skylightVisualNotifications-tenantId-agentId",
  ],
]
`;

exports[`goToggleAudioNotificationsPreference gets the current audio notification 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleAudioNotificationsPreference puts the inverse of it 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "audioNotifications": false,
        "type": "app/AgentNotificationsMenu/SET_AUDIO_NOTIFICATIONS_PREFERENCE",
      },
      "channel": null,
    },
  },
}
`;

exports[`goToggleAudioNotificationsPreference selectes the agent id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleAudioNotificationsPreference selectes the tenant id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleAudioNotificationsPreference uses the tenant and agent id to set the notification in localStorage 1`] = `
Array [
  Array [
    "skylightAudioNotifications-tenantId-agentId",
    false,
  ],
]
`;

exports[`goToggleVisualNotificationsPreference does have notification permission gets the current visual notification 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does have notification permission puts the inverse of it 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/AgentNotificationsMenu/SET_VISUAL_NOTIFICATIONS_PREFERENCE",
        "visualNotifications": false,
      },
      "channel": null,
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does have notification permission selectes the agent id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does have notification permission selectes the tenant id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does have notification permission uses the tenant and agent id to set the notification in localStorage 1`] = `
Array [
  Array [
    "skylightVisualNotifications-tenantId-agentId",
    false,
  ],
]
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is granted upon request requests permission 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is granted upon request selectes the agent id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is granted upon request selectes the tenant id 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "SELECT": Object {
      "args": Array [],
      "selector": [Function],
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is granted upon request sets the notification to true 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/AgentNotificationsMenu/SET_VISUAL_NOTIFICATIONS_PREFERENCE",
        "visualNotifications": true,
      },
      "channel": null,
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is granted upon request uses the tenant and agent id to set the notification in localStorage 1`] = `
Array [
  Array [
    "skylightVisualNotifications-tenantId-agentId",
    true,
  ],
]
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is not granted upon request adds error 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "error": Object {
          "code": "AD-1006",
        },
        "interactionFatal": undefined,
        "type": "app/Errors/SET_NON_CRITICAL_ERROR",
      },
      "channel": null,
    },
  },
}
`;

exports[`goToggleVisualNotificationsPreference does not have notification permission permission is not granted upon request requests permission 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;
